waveform_editor/dependency_graph.py

```diff
@@ -0,0 +1,65 @@
+from waveform_editor.derived_waveform import DerivedWaveform
+
+
+class DependencyGraph:
+    def __init__(self):
+        self.graph = {}
+
+    def add_node(self, name, dependencies):
+        self.graph[name] = set(dependencies)
+        try:
+            self.detect_cycles()
+        except RuntimeError as e:
+            del self.graph[name]
+            raise e
+
+    def remove_node(self, name):
+        if name in self.graph:
+            del self.graph[name]
+        for deps in self.graph.values():
+            deps.discard(name)
```

Comment:
We shouldn't do this, right?
The underlying waveform still has a dependency on `name`.